### PR TITLE
Add service command

### PR DESCRIPTION
**Pull request by kirtangajjar**
Issue ref - https://github.com/EasyEngine/easyengine/issues/1010

closes - https://github.com/EasyEngine/easyengine/issues/1010

### CODE REVIEW

File: src/Service_Command.php [hunk 1]
```diff
@@ -15,32 +15,33 @@
 
 class Service_Command extends EE_Command {
 
+	private $global_container_name = 'ee-nginx-proxy';
 	/**
 	 * Starts global reverse proxy container.
 	 */
 	public function start( $cmd, $descriptors = null ) {
-		\EE\Utils\default_launch( "docker start ee-nginx-proxy" );
+		\EE\Utils\default_launch( "docker start $this->container_name" );
```
Comment (mrrobot47) [inline src/Service_Command.php]: Should it not be `$this->global_container_name`

Comment (mrrobot47) [reply src/Service_Command.php]: Also other files like `composer.json` etc are missing. Please add all of them.

**Comment by mrrobot47** [comment]
LGTM :+1:

File: src/Service_Command.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+/**
+ * Executes wp-cli command on a site.
```
Comment (rahulsprajapati) [inline src/Service_Command.php]: I think this comment is not related to functionality added to this file.

File: src/Service_Command.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @package ee-cli
+ */
+
+use EE\Utils;
+
+class Service_Command extends EE_Command {
+
+	private $global_container_name = 'ee-nginx-proxy';
+	/**
```
Comment (rahulsprajapati) [inline src/Service_Command.php]: Keep a new line above this.

**Comment by kirtangajjar** [comment]
Made changes as requested

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
+{
+    "name": "easyengine/service-command",
+    "description": "Shell to run helpful commands inside containers.",
```
Comment (mbtamuli) [inline composer.json]: This description is of the shell-command I believe.

File: README.md [hunk 5]
```diff
@@ -1 +1,19 @@
...
+
+So let's say you have foo.com and bar.com on same ee server. When anyone requests either website, the request will first go to `ee-nginx-proxy` which will forward the request to appropriate nginx container of site (In this case foo.com or bar.com).
+
+## Usage:
+
+```
+ee service start    # starts global container
+ee service restart    # restarts global container
+ee service stop    # stops global container
+ee service reload    # reloads global container
```
Comment (mbtamuli) [inline README.md]:
```
ee service reload             Reload the configuration of the ee-nginx-proxy service
```

Comment (kirtangajjar) [reply README.md]: @mbtamuli I'm thinking about `Reloads the configuration of the global service` since it can contain multiple global containers in future.

File: composer.json (hunk 4 above)
Comment (kirtangajjar) [reply composer.json]: Yes. Fixed

File: README.md (hunk 5 above)
Comment (mbtamuli) [reply README.md]: > multiple global containers in future.

It will have to something like 
```
ee service reload --ee-nginx-proxy
ee service reload --ee-mysql
ee service reload --all
```

Comment (kirtangajjar) [reply README.md]: @mbtamuli Done

File: src/Service_Command.php (hunk 3 above)
Comment (kirtangajjar) [reply src/Service_Command.php]: @rahulsprajapati Done.

File: src/Service_Command.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ *     $ ee service restart nginx-proxy
+ *
+ * @package ee-cli
+ */
+
+use EE\Utils;
+
+class Service_Command extends EE_Command {
+
+	private $whitelisted_containers = [
```
Comment (rahulsprajapati) [inline src/Service_Command.php]: comments?

File: src/Service_Command.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		$containers = array_intersect( $this->whitelisted_containers, $args );
+
+		if( empty( $containers ) ) {
+			EE::error( "Unable to find global EasyEngine container $args[0]" );
+		}
+
+		return $containers[0];
+	}
+
+	private function container_reload_command( $container ) {
```
Comment (rahulsprajapati) [inline src/Service_Command.php]: comment?

**Comment by kirtangajjar** [comment]
@rahulsprajapati Fixed requested changes

File: src/Service_Command.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+ *     $ ee service restart nginx-proxy
+ *
+ * @package ee-cli
+ */
+class Service_Command extends EE_Command {
+
+	/**
+	 * @var array Array of containers defined in global docker-compose.yml
+	 */
+	private $whitelisted_containers = [
```
Comment (mbtamuli) [inline src/Service_Command.php]: Although we are interacting with containers, we are not directly interacting with the containers themselves.
As such you've not used the container name but the service name mentioned in docker-compose.yml.

So although there is no functional change, but I think it would be better if we renamed almost all instances of the word *container* with *service*.

File: src/Service_Command.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Starts global containers.
+	 *
+	 * ## OPTIONS
+	 *
+	 * <service-name>
+	 * : Name of container.
+	 */
+	public function start( $args, $assoc_args ) {
+		$container = $this->filter_container( $args );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+		EE::exec( "docker-compose start $container", true, true );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	/**
+	 * Returns valid container name from arguments.
+	 */
+	private function filter_container( $args ) {
+		$containers = array_intersect( $this->whitelisted_containers, $args );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	public function start( $args, $assoc_args ) {
+		$container = $this->filter_container( $args );
+
+		EE::exec( "docker-compose start $container", true, true );
+	}
+
+	/**
+	 * Returns valid container name from arguments.
+	 */
+	private function filter_container( $args ) {
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+		$containers = array_intersect( $this->whitelisted_containers, $args );
+
+		if ( empty( $containers ) ) {
+			EE::error( "Unable to find global EasyEngine container $args[0]" );
+		}
+
+		return $containers[0];
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+		EE::exec( "docker-compose start $container", true, true );
+	}
+
+	/**
+	 * Returns valid container name from arguments.
+	 */
+	private function filter_container( $args ) {
+		$containers = array_intersect( $this->whitelisted_containers, $args );
+
+		if ( empty( $containers ) ) {
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Stops global containers.
+	 *
+	 * ## OPTIONS
+	 *
+	 * <service-name>
+	 * : Name of container.
+	 */
+	public function stop( $args, $assoc_args ) {
+		$container = $this->filter_container( $args );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+		EE::exec( "docker-compose stop $container", true, true );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	/**
+	 * Restarts global containers.
+	 *
+	 * ## OPTIONS
+	 *
+	 * <service-name>
+	 * : Name of container.
+	 */
+	public function restart( $args, $assoc_args ) {
+		$container = $this->filter_container( $args );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+		EE::exec( "docker-compose restart $container", true, true );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+	 *
+	 * ## OPTIONS
+	 *
+	 * <service-name>
+	 * : Name of container.
+	 */
+	public function reload( $args, $assoc_args ) {
+		$container = $this->filter_container( $args );
+		$command   = $this->container_reload_command( $container );
+		EE::exec( "docker-compose exec $container $command", true, true );
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

File: src/Service_Command.php [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+	 * @param $container name of conntainer
+	 *
+	 * @return mixed
+	 */
+	private function container_reload_command( string $container ) {
+		$command_map = [
+			'nginx-proxy' => "sh -c 'nginx -t && service nginx reload'",
+		];
+
+		return $command_map[ $container ];
```
Comment (mbtamuli) [inline src/Service_Command.php]: Replace the word *container* with *service*

**Comment by kirtangajjar** [comment]
@mbtamuli Fixed.
